### DISCUSSION TITLE
pnpm monorepo support

### DISCUSSION THREAD

**Discussion by acherkashin** [sandworm-hq/sandworm-audit#101] [Ideas]
**Sandworm version**
1.29.1

**Describe the bug**
Does sandworm-audit supports pnpm monorepos?

In the root of my repository there are 2 files:
- pnpm-lock.yaml
- pnpm-workspace.yaml
 
<img width="376" alt="image" src="https://user-images.githubusercontent.com/9947582/224901019-daae6a0a-1c3b-406d-8aca-a0cc6cc124ae.png">

There is no `package.json` in the same folder with `pnpm-lock.yaml`. At the same time, every package of monorepo has a `package.json`.

When I try to run `sandworm-audit` in the root of the repository, I get the following error.

```
TypeError: Cannot read properties of undefined (reading 'name')
    at generatePnpmGraph (/Users/cherkalexander/.nvm/versions/node/v16.15.1/lib/node_modules/@sandworm/audit/src/graph/generatePnpmGraph.js:15:20)
    at generateGraphPromise (/Users/cherkalexander/.nvm/versions/node/v16.15.1/lib/node_modules/@sandworm/audit/src/graph/index.js:37:19)
    at async getReport (/Users/cherkalexander/.nvm/versions/node/v16.15.1/lib/node_modules/@sandworm/audit/src/index.js:41:25)
    at async Object.handler (/Users/cherkalexander/.nvm/versions/node/v16.15.1/lib/node_modules/@sandworm/audit/src/cli/index.js:43:7)
```

**System (please complete the following information):**
- Node Version - 16.15.1
- Package Manager - PNPM 
- Package Manager Version - 7.17.1
- OS: MacOs

**Additional context**
I tried to use sandworm in a project where yarn 1 is used, and it works great. Thank you for this cool project 👍

**Comment by gabidobo** [comment]
Hi @acherkashin!

Thank you for the kind words 😊 

You are correct - Sandworm doesn't currently support workspaces. I've added some notes in the docs to make this explicit.
Workspaces are definitely on the roadmap, I just need to understand how that feature would look like. If you have any suggestions they are welcome - as a user, how would you like to use Sandworm to audit workspaces?

**Reply by jnurkka** [reply]
We'd also love this feature! We already use sandworm in a yarn v1 workspace. We didn't face any issues there, I'm assuming because the yarn v1 lockfile is supported by sandworm?

I'm not sure what you mean with how we'd like to use Sandworm to audit workspaces, but basically what we'd like to do is just run sandworm audit against the `pnpm-lock.yaml` in our workspace root in our CI.

If you need any further info, happy to provide that!

**Reply by liamjones** [reply]
In our case, I'd like to use it at two levels;
1. At the monorepo root to check nothing licensed in a way that will cause us issues from a commercial licensing point of view
2. At an individual project level within the monorepo to help with license attribution info for an individual published artifact (RN apps in our case)

**Reply by gabidobo** [reply]
Thank you all for the feedback! We've released workspace support in v1.45.0 🎉 

You'll need to run `sandworm-audit` in each individual package directory you want to audit instead of the workspace root.

Let me know if it works for you as well 👍

**Reply by acherkashin** [reply]
@gabidobo it is good news

I tried to run it on my production project, but sandworm just stuck on `Getting vulnerability report from pnpm` stage for some reason 🤔 

<img width="379" alt="image" src="https://github.com/sandworm-hq/sandworm-audit/assets/9947582/304e20b7-73c2-4302-801a-9651f42fae2e">

**Reply by liamjones** [reply]
@gabidobo FYI, still not working here. I've tried running from both the root with `-p` and from an app package subdirectory itself and it still doesn't seem to be finding the `yarn.lock` in the repo root:

```shell
monorepo on  develop [$⇡]
❯ sandworm-audit -p packages/apps/an-app
Sandworm 🪱
Security and License Compliance Audit
⠋ Building dependency graph...
❌ Failed: No lockfile found

monorepo on  develop [$⇡]
❯ cd packages/apps/an-app

packages/apps/an-app on  develop [$⇡]
❯ sandworm-audit
Sandworm 🪱
Security and License Compliance Audit
⠋ Building dependency graph...
❌ Failed: No lockfile found

packages/apps/an-app on  develop [$⇡]
❯ sandworm-audit -v
1.46.0
```

**Reply by gabidobo** [reply]
Thank you so much for taking the time to test this out!

I've opened a couple of issues based on the reports above, will follow-up there:

- #130 
- #131
